File: barnyard/effects.py
"""Particles, floating score popups and screen shake."""

from __future__ import annotations

import math
import random
from dataclasses import dataclass

import pygame

from . import config


@dataclass
class Particle:
    x: float
    y: float
    vx: float
    vy: float
    life: float
    max_life: float
    size: float
    color: tuple
    spin: float = 0.0
    angle: float = 0.0


@dataclass
class Popup:
    x: float
    y: float
    text: str
    color: tuple
    life: float = 0.9
    max_life: float = 0.9
    size: int = 26


class Effects:
    def __init__(self, rng: random.Random | None = None):
        self.rng = rng or random.Random()
        self.particles: list[Particle] = []
        self.popups: list[Popup] = []
        self.shake = 0.0

    # ------------------------------------------------------------------ spawn
    def burst(self, x: float, y: float, color, count: int = 12,
              speed: float = 260.0) -> None:
        for _ in range(count):
            angle = self.rng.uniform(0, math.tau)
            mag = self.rng.uniform(0.35, 1.0) * speed
            self.particles.append(Particle(
                x, y,
                math.cos(angle) * mag, math.sin(angle) * mag - 60,
                life=self.rng.uniform(0.35, 0.7),
                max_life=0.7,
                size=self.rng.uniform(3, 7),
                color=color,
                spin=self.rng.uniform(-8, 8),
            ))

    def ring(self, x: float, y: float, color, count: int = 22,
             speed: float = 420.0) -> None:
        for i in range(count):
            angle = math.tau * i / count
            self.particles.append(Particle(
                x, y,
                math.cos(angle) * speed, math.sin(angle) * speed,
                life=0.45, max_life=0.45,
                size=self.rng.uniform(4, 8),
                color=color,
                spin=self.rng.uniform(-6, 6),
            ))

    def feathers(self, x: float, y: float, count: int = 16) -> None:
        palette = [(255, 250, 235), (250, 226, 150), (240, 158, 42)]
        for _ in range(count):
            self.particles.append(Particle(
                x, y,
                self.rng.uniform(-160, 160), self.rng.uniform(-260, -60),
                life=self.rng.uniform(0.6, 1.1), max_life=1.1,
                size=self.rng.uniform(4, 8),
                color=self.rng.choice(palette),
                spin=self.rng.uniform(-10, 10),
            ))

    def popup(self, x: float, y: float, text: str, color=config.CREAM,
              size: int = 26) -> None:
        self.popups.append(Popup(x, y, text, color, size=size))

    def kick(self, amount: float) -> None:
        self.shake = min(14.0, self.shake + amount)

    # ----------------------------------------------------------------- update
    def update(self, dt: float) -> None:
        alive: list[Particle] = []
        for p in self.particles:
            p.life -= dt
            if p.life <= 0:
                continue
            p.vy += 900 * dt
            p.vx *= 0.98
            p.x += p.vx * dt
            p.y += p.vy * dt
            p.angle += p.spin * dt
            alive.append(p)
        self.particles = alive

        live_popups: list[Popup] = []
        for pop in self.popups:
            pop.life -= dt
            if pop.life > 0:
                pop.y -= 46 * dt
                live_popups.append(pop)
        self.popups = live_popups

        self.shake = max(0.0, self.shake - 44 * dt)

    def offset(self) -> tuple[int, int]:
        if self.shake <= 0.1:
            return (0, 0)
        return (self.rng.randint(-1, 1) * int(self.shake),
                self.rng.randint(-1, 1) * int(self.shake))

    # ------------------------------------------------------------------- draw
    def draw(self, surface: pygame.Surface, font_for) -> None:
        for p in self.particles:
            alpha = max(0.0, min(1.0, p.life / p.max_life))
            size = max(2, int(p.size * (0.4 + 0.6 * alpha)))
            chip = pygame.Surface((size * 2, size * 2), pygame.SRCALPHA)
            pygame.draw.rect(chip, (*p.color, int(255 * alpha)),
                             chip.get_rect(), border_radius=size // 2)
            chip = pygame.transform.rotate(chip, math.degrees(p.angle))
            surface.blit(chip, chip.get_rect(center=(int(p.x), int(p.y))))

        for pop in self.popups:
            t = pop.life / pop.max_life
            font = font_for(pop.size)
            alpha = int(255 * min(1.0, t * 2.2))
            shadow = font.render(pop.text, True, (40, 30, 26))
            label = font.render(pop.text, True, pop.color)
            shadow.set_alpha(int(alpha * 0.5))
            label.set_alpha(alpha)
            rect = label.get_rect(center=(int(pop.x), int(pop.y)))
            surface.blit(shadow, rect.move(2, 2))
            surface.blit(label, rect)

File: app/src/main/kotlin/com/barnyardblitz/ui/Effects.kt
package com.barnyardblitz.ui

import android.graphics.Canvas
import android.graphics.RectF
import kotlin.math.abs
import kotlin.math.roundToInt
import kotlin.random.Random

/** Particles, floating score popups and screen shake. */
class Effects(private val random: Random = Random.Default) {

    private class Particle(
        var x: Float, var y: Float, var vx: Float, var vy: Float,
        var life: Float, val maxLife: Float, val size: Float, val color: Int,
    )

    private class Popup(
        var x: Float, var y: Float, val text: String, val color: Int,
        var life: Float, val maxLife: Float, val size: Float,
    )

    private val particles = ArrayList<Particle>()
    private val popups = ArrayList<Popup>()
    var shake: Float = 0f
        private set

    fun clear() {
        particles.clear()
        popups.clear()
        shake = 0f
    }

    fun burst(x: Float, y: Float, color: Int, count: Int = 12, speed: Float = 260f) {
        repeat(count) {
            val angle = random.nextDouble(0.0, Math.PI * 2).toFloat()
            val mag = (0.35f + random.nextFloat() * 0.65f) * speed
            particles.add(
                Particle(
                    x, y,
                    Math.cos(angle.toDouble()).toFloat() * mag,
                    Math.sin(angle.toDouble()).toFloat() * mag - 60f,
                    0.35f + random.nextFloat() * 0.35f, 0.7f,
                    3f + random.nextFloat() * 4f, color,
                ),
            )
        }
    }

    fun ring(x: Float, y: Float, color: Int, count: Int = 22, speed: Float = 420f) {
        for (i in 0 until count) {
            val angle = (Math.PI * 2 * i / count).toFloat()
            particles.add(
                Particle(
                    x, y,
                    Math.cos(angle.toDouble()).toFloat() * speed,
                    Math.sin(angle.toDouble()).toFloat() * speed,
                    0.45f, 0.45f, 4f + random.nextFloat() * 4f, color,
                ),
            )
        }
    }

    fun feathers(x: Float, y: Float, count: Int = 16) {
        val palette = intArrayOf(0xFFFFFAEB.toInt(), 0xFFFAE296.toInt(), 0xFFF09E2A.toInt())
        repeat(count) {
            particles.add(
                Particle(
                    x, y,
                    -160f + random.nextFloat() * 320f, -260f + random.nextFloat() * 200f,
                    0.6f + random.nextFloat() * 0.5f, 1.1f,
                    4f + random.nextFloat() * 4f, palette[random.nextInt(palette.size)],
                ),
            )
        }
    }

    fun popup(x: Float, y: Float, text: String, color: Int, size: Float) {
        popups.add(Popup(x, y, text, color, 0.9f, 0.9f, size))
        // A short stack: past three or four, labels just overlap each other.
        while (popups.size > 4) popups.removeAt(0)
    }

    fun kick(amount: Float) {
        shake = minOf(14f, shake + amount)
    }

    fun update(dt: Float) {
        var i = particles.size - 1
        while (i >= 0) {
            val particle = particles[i]
            particle.life -= dt
            if (particle.life <= 0f) {
                particles.removeAt(i)
            } else {
                particle.vy += 900f * dt
                particle.vx *= 0.98f
                particle.x += particle.vx * dt
                particle.y += particle.vy * dt
            }
            i--
        }
        var j = popups.size - 1
        while (j >= 0) {
            val popup = popups[j]
            popup.life -= dt
            if (popup.life <= 0f) popups.removeAt(j) else popup.y -= 46f * dt
            j--
        }
        shake = maxOf(0f, shake - 44f * dt)
    }

    fun offsetX(): Float = if (shake <= 0.1f) 0f else (random.nextInt(3) - 1) * shake
    fun offsetY(): Float = if (shake <= 0.1f) 0f else (random.nextInt(3) - 1) * shake

    fun draw(canvas: Canvas, ui: Ui) {
        for (particle in particles) {
            val alpha = (particle.life / particle.maxLife).coerceIn(0f, 1f)
            val size = maxOf(2f, particle.size * (0.4f + 0.6f * alpha))
            ui.fill.color = withAlpha(particle.color, alpha)
            canvas.drawRoundRect(
                RectF(particle.x - size, particle.y - size, particle.x + size, particle.y + size),
                size / 2f, size / 2f, ui.fill,
            )
        }
        for (popup in popups) {
            val t = popup.life / popup.maxLife
            val alpha = minOf(1f, t * 2.2f)
            ui.text(
                canvas, popup.text, popup.size, withAlpha(popup.color, alpha),
                popup.x, popup.y, Ui.Align.CENTER, bold = true,
            )
        }
    }

    /** Used by tests and the view to know whether a redraw is still needed. */
    val busy: Boolean get() = particles.isNotEmpty() || popups.isNotEmpty() || shake > 0.1f
}

/** A short message that slides up from the button bar. */
class Toast(val text: String, val color: Int) {
    var life: Float = 2.4f
    val maxLife: Float = 2.4f
}

/** Shared by every scene. */
interface Scene {
    fun onEnter(argument: String? = null) {}
    fun onLayout() {}
    fun update(dt: Float) {}
    fun draw(canvas: Canvas)
    fun onDown(x: Float, y: Float) {}
    fun onMove(x: Float, y: Float) {}
    fun onUp(x: Float, y: Float) {}
    /** Return true when the scene consumed the back gesture. */
    fun onBack(): Boolean = false
}

internal fun easeOut(t: Float): Float {
    val clamped = t.coerceIn(0f, 1f)
    val inv = 1f - clamped
    return 1f - inv * inv * inv
}

internal fun easeIn(t: Float): Float = t.coerceIn(0f, 1f).let { it * it }

internal fun formatCoins(value: Int): String {
    if (abs(value) < 1000) return value.toString()
    val text = value.toString()
    val out = StringBuilder()
    var count = 0
    for (i in text.length - 1 downTo 0) {
        out.append(text[i])
        count++
        if (count % 3 == 0 && i > 0 && text[i - 1].isDigit()) out.append(',')
    }
    return out.reverse().toString()
}

internal fun Float.roundToIntSafe(): Int = if (isNaN()) 0 else roundToInt()
